examples/client_middleware_auth.py

```diff
@@ -0,0 +1,525 @@
+        cls, auth_url: str, username: str, password: str, session: ClientSession
+    ) -> "TokenAuthMiddleware":
+        """Factory method to create middleware with a session and authenticate immediately."""
+        middleware = cls(auth_url, username, password)
+        middleware.set_session(session)
+        await middleware.authenticate()
+        return middleware
+
+    async def authenticate(self) -> None:
+        """Get initial authentication tokens."""
+        if not self.session:
+            raise RuntimeError("Session not set. Call set_session() first.")
+
+        async with self.session.post(
+            self.auth_url,
+            json={"username": self.username, "password": self.password},
+            # Don't use middlewares here to avoid recursion
+            middlewares=(),
+        ) as resp:
+            data: Dict[str, str] = await resp.json()
+            self.access_token = data["access_token"]
+            self.refresh_token = data.get("refresh_token")
+            _LOGGER.info("Authenticated as %s", self.username)
+
+    async def refresh_access_token(self) -> None:
+        """Refresh the access token using refresh token."""
+        if not self.session:
+            raise RuntimeError("Session not set. Call set_session() first.")
+
+        if not self.refresh_token:
+            # No refresh token, need to re-authenticate
+            await self.authenticate()
+            return
+
+        async with self.session.post(
+            f"{self.auth_url}/refresh",
+            json={"refresh_token": self.refresh_token},
+            # Don't use middlewares here to avoid recursion
+            middlewares=(),
+        ) as resp:
+            if resp.status != 200:
+                # Refresh failed, re-authenticate
+                _LOGGER.warning("Token refresh failed, re-authenticating")
+                await self.authenticate()
+                return
+
+            data: Dict[str, str] = await resp.json()
+            self.access_token = data["access_token"]
+            _LOGGER.info("Token refreshed successfully")
+
+    async def middleware(
+        self, request: ClientRequest, handler: ClientHandlerType
+    ) -> ClientResponse:
+        """Apply authentication to requests and handle token refresh."""
+        # Skip auth endpoints
+        if str(request.url).startswith(self.auth_url):
+            return await handler(request)
+
+        retry_with_refresh = False
+
+        while True:
+            # Add auth header if we have a token
+            if self.access_token:
+                request.headers["Authorization"] = f"Bearer {self.access_token}"
+
+            # Make the request
+            response = await handler(request)
+
+            # Check if token expired
+            if response.status == 401 and not retry_with_refresh:
+                _LOGGER.info("Got 401, attempting to refresh token")
+                # Try to refresh the token
+                await self.refresh_access_token()
```

Comment:
Can we verify this also works by just passing request.session rather than needing a reference to the session beforehand (we'll need to add a new property method)?

Reply:
https://github.com/aio-libs/aiohttp/pull/9732/commits/4417d9c8f8a4e5fdb07f6bb4bfdf7e4260d9e13c